### PR TITLE
python: add mesonpep517 as build system

### PR DESCRIPTION
**Pull request by FRidh**
###### Motivation for this change

I would like to demonstrate mesonpep517 and thus choose to upstream it.

While it "works", I am not very happy with this yet. See e.g. the comments in the code.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/development/python-modules/mesonpep517/default.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  # Meson tries to detect ninja as well, so we should patch meson as well.
+  # postPatch = ''
+  #   substituteInPlace mesonpep517/buildapi.py \
+  #     --replace "'meson'" "'${meson}/bin/meson'" \
+  #     --replace "'ninja'" "'${ninja}/bin/ninja'"      
+  # '';
+
+  propagatedNativeBuildInputs = [ meson ninja ];
+
+  # This should be propagated as long as meson comes with a hook.
```
Comment (FRidh) [inline pkgs/development/python-modules/mesonpep517/default.nix]: I am starting to get an increasing dislike of packages *including* a setup hook when you do not want it. Maybe we should split it off, that is into a `meson` and `mesonHook`.

File: pkgs/development/python-modules/mesonpep517/default.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  # Unless we patch mesonpep517 to a meson path?
+  dontUseMesonConfigure = true;
+
+  meta = {
+    description = "Create pep517 compliant packages from the meson build system";
+    homepage = https://gitlab.com/thiblahute/mesonpep517;
+    license = lib.licenses.asl;
```
Comment (FRidh) [inline pkgs/development/python-modules/mesonpep517/default.nix]: incorrect license

**Comment by jtojnar** [comment]
We will need to patch this to support multiple outputs:

https://gitlab.com/thiblahute/mesonpep517/blob/d3ca51c80a7960b71b2f8852c235d9269cf2e038/mesonpep517/buildapi.py#L44-51

File: pkgs/development/python-modules/mesonpep517/default.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+, buildPythonPackage
+, fetchPypi
+, meson
+, ninja
+, intreehooks
+, pytoml
+, pythonOlder
+}:
+
+# TODO: offer meson as a Python package so we have dist-info folder.
```
Comment (FRidh) [inline pkgs/development/python-modules/mesonpep517/default.nix]: It's silly, actually. We need it is as a Python package if packages start adding it as a build system requirement. However, it should not be needed considering meson is only an *application*; they do not actually import it.

**Comment by FRidh** [comment]
@grahamcofborg build python3.pkgs.mesonpep517

**Pull request merged**